Generic Drake/ROS systems + RS Flip Flop example

sloretz opened the pull request:
This adds generic ROS publisher/subscriber systems as described in robotlocomotion/drake#9500. It's similar to https://github.com/gizatt/drake_ros_systems , but this uses ROS 2 and matches the LCM publisher/subscriber system's design more closely.

# RS Flip Flop example

There's an example of an RS flip flop system that listens to topics `/S`, and `/R` and publishes to topics `/Q` and `/Q_not`.

To try this example out:
1. Install ROS Rolling - it might work with other ROS 2 versions, but I haven't tried it
1. Clone this repo into a colcon workspace
1. `colcon build --packages-select drake_ros_systems`
1. Run one of the `rs_flip_flop` examples
   * C++ from the `build` folder: `./build/drake_ros_systems/example/rs_flip_flop`
   * Python from the `example` folder: `python3 rs_flip_flop.py`
1. Play with it with some of these commands:

Echo output topics:
```
$ ros2 topic echo /Q
```

```
ros2 topic echo /Q_not
```

Send values to the input topics

```
$ ros2 topic pub /S std_msgs/msg/Bool "data: false"
$ ros2 topic pub /S std_msgs/msg/Bool "data: true"
```

```
$ ros2 topic pub /R std_msgs/msg/Bool "data: false"
$ ros2 topic pub /R std_msgs/msg/Bool "data: true"
```

# Future enhancements

C++

* Allow picking the executor and executor options in `DrakeRos`
* Add return value to `spin` functions in `rclcpp` that indicate f if spin did something, and if so pause Drake's concept of time to handle multiple messages [just like LcmInterfaceSystem does](https://github.com/RobotLocomotion/drake/blob/abfb13c48b052cf5a69ebbb707f64432b63668c6/systems/lcm/lcm_interface_system.cc#L63-L71).
* Replace Publisher and Subscriber classes with Generic Publisher and Subscriber being worked on in https://github.com/ros2/rclcpp/pull/1452

Python
* Use `pybind11::capsule::get_pointer()` when available

# Code copying

The generic publisher and subscriber using raw messages was significantly copied from `rosbag2_transport`: https://github.com/ros2/rosbag2/tree/master/rosbag2_transport but it should probably be change to use these classes (not yet merged https://github.com/ros2/rclcpp/pull/1452)

The Drake boilerplate in RosInterfaceSystem, RosSubscriberSystem, and RosPublisherSystem was significantly copied from the equivalent LCM systems in drake.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sloretz/drake_ros2_demos/6)
<!-- Reviewable:end -->

sloretz committed:
Stub code builds

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Misc fixes to get it to build (but not link) with stub example

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix comments on publisher system

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
ROS interface implementation fleshed out

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Possible RosInterfaceSystem impl

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Possible RosSubscriberSystem impl

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Possible RosPublisherSystem implementation

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
RosPublisherSystem uses separate Serializer interface

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
RosSubscriberSystem uses separate Serializer interface

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add RS-flip-flop example

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented on drake_ros_systems/README.md:
```diff
@@ -0,0 +1,62 @@
+# Drake ROS Systems
```
If reviewing, I recommend skipping the README for now. I haven't updated it since I got the example working.

sloretz replied:
The README is now useful :tada:

sloretz commented:
@IanTheEngineer @EricCousineau-TRI @azeey FYI

sloretz committed:
Add get_type_support() to SerializerInterface

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

EricCousineau-TRI commented on drake_ros_systems/include/drake_ros_systems/drake_ros.hpp:
```diff
@@ -0,0 +1,52 @@
+#ifndef DRAKE_ROS_SYSTEMS__DRAKE_ROS_HPP_
```
Should this be called `ros2` to clearly delineate?

(We may want to incorporate `ros1` support based on academic collaboration, after we prove out ROS2 support in Anzu.)

sloretz replied:
I would recommend keeping the name agnostic to ROS 1 and ROS 2. For Drake users the name won't make much of a difference. That will let them migrate from one to the other without changing the includes to `drake_ros_system`, or really much of how they interact with the ROS messages themselves. I think they'd just have to change the C++ ROS message `#includes` to include `/msg/` and the `QoS` options to the system constructors.

For the ROS side, there are a few examples of suffixing the version with the name has causing some maintenance headaches in the ROS ecosystem (`ros/geometry` to `ros/geometry2`, the duplicates on the `ros2` org like `ros/kdl_parser` -> [`ros2/kdl_parser`](https://github.com/ros2/kdl_parser)).

An alternative is supporting ROS 1 and ROS 2 on different branches in the same repo (Now done on repos like [`kdl_parser`](https://github.com/ros/kdl_parser)), or supporting both ROS 1 and ROS 2 at the same time (done by some repos like [PlotJuggler](https://github.com/facontidavide/PlotJuggler)).

EricCousineau-TRI replied:
Hm... The main concern is that ros1 and ros2 packages may then become unusable in the same binary.

Isn't it possible to consume them both in the same binary? (can ament consume catkin, and can ROS 1 libs avoid stepping on the feet of ROS 2 libs for symbols?)

EricCousineau-TRI replied:
Ah, sorry, just read your part about the suffixing. Yeah, I think the forking of rviz2 is also part of that (though it was refactored).

OK, I'm down for trying out the forking approach first, and then pivoting later if it actually becomes a problem.
Thanks!

sloretz committed:
Hello world pub example

Publisher and ROS interface working, subscriber not yet working.

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix includes for DrakeRosInterface and RosInterfaceSystem

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
ROS Subscriber works in Python

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Implement flip-flop example in Python

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Update README for C++ or Python example

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Try more whitespace

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Bullet to numbered list

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Commands to play with the example

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

azeey reviewed (approved):
LGTM! However, I don't understand much of the code in `py_serializer.hpp`, so another pair of eyes would be welcome.

I have built and tested the example with both the colcon and cmake workflows, and they both work!

sloretz committed:
Blank line at EOF

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add linter tests

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add LICENSE file and header comments

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unused String import

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Satisfy uncrustify

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Satisfy cpplint

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Custom node name, node opts, and init opts

This adds a way to provide the node name and node options, as well as a
way to externally manage init/shutdown on a context. If the given node
options have a valid context, then DrakeRos won't call init or shutdown
on it. In this way external code can pass init options.

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add additional publish triggers

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Subscriber already unsubs on destruction

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Allow setting QoS on Python systems

Adds half a type caster (Python -> C++) for rclpy.qos.QoSProfile to
rclcpp::QoS.
Adds QoS arguments to Python publisher and subscriber systems.
Updates the Python example to pass QoS.

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Expose publish triggers to Python systems

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Throw exception if Serialization fails

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove obsolete comment

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

EricCousineau-TRI commented:
> LGTM! However, I don't understand much of the code in py_serializer.hpp, so another pair of eyes would be welcome.

Placing one pair of :eyes: here

EricCousineau-TRI commented:
(Going to use Reviewable tho)

EricCousineau-TRI reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sloretz/drake_ros2_demos/6)* status: 0 of 25 files reviewed, 3 unresolved discussions (waiting on @sloretz)

___
*[drake_ros_systems/src/python_bindings/py_serializer.hpp, line 30 at r4](https://reviewable.io/reviews/sloretz/drake_ros2_demos/6#-MOlpl5g3qsc2CmTWrBH:-MOlpl5g3qsc2CmTWrBI:b32ctja) ([raw file](https://github.com/sloretz/drake_ros2_demos/blob/41d8f8a30f731875e06476b8291fd58d4b7f95f5/drake_ros_systems/src/python_bindings/py_serializer.hpp#L30)):*
> ```cpp
> {
> // Serialize/Deserialize Python ROS types to rclcpp::SerializedMessage
> class PySerializer : public SerializerInterface
> ```

It's a bit challenging to analyze and review Python code flow in C++.
(Also, pumping stuff back and forth btw the two may be less efficient?)

As my own rule of thumb, anything that calls `pybind11`s Python C++ API more than 3 times in succession should generally be done in a Python function.

Is it possible for you to relegate this code to Python, and use something like what Drake to does to invoke Python code?

Example:
https://github.com/RobotLocomotion/drake/blob/2baad6f34b12bfea1c583dfd63b0970b7eeb479a/bindings/pydrake/pydrake_pybind.h#L127

https://github.com/RobotLocomotion/drake/blob/2baad6f34b12bfea1c583dfd63b0970b7eeb479a/bindings/pydrake/__init__.py#L77-L80


<!-- Sent from Reviewable.io -->

sloretz committed:
NodeOptions defaults to global context

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add integration test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add Python integration test

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove debug print

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix bug where context wasn't being shut down

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add tests for DrakeRos

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Do more in Python

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sloretz/drake_ros2_demos/6)* status: 0 of 27 files reviewed, 3 unresolved discussions (waiting on @EricCousineau-TRI and @sloretz)

___
*[drake_ros_systems/src/python_bindings/py_serializer.hpp, line 30 at r4](https://reviewable.io/reviews/sloretz/drake_ros2_demos/6#-MOlpl5g3qsc2CmTWrBH:-MOryhoK1y-HCrWHeS-Z:b-aiwpuc) ([raw file](https://github.com/sloretz/drake_ros2_demos/blob/41d8f8a30f731875e06476b8291fd58d4b7f95f5/drake_ros_systems/src/python_bindings/py_serializer.hpp#L30)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

It's a bit challenging to analyze and review Python code flow in C++.
(Also, pumping stuff back and forth btw the two may be less efficient?)

As my own rule of thumb, anything that calls `pybind11`s Python C++ API more than 3 times in succession should generally be done in a Python function.

Is it possible for you to relegate this code to Python, and use something like what Drake to does to invoke Python code?

Example:
https://github.com/RobotLocomotion/drake/blob/2baad6f34b12bfea1c583dfd63b0970b7eeb479a/bindings/pydrake/pydrake_pybind.h#L127

https://github.com/RobotLocomotion/drake/blob/2baad6f34b12bfea1c583dfd63b0970b7eeb479a/bindings/pydrake/__init__.py#L77-L80
</blockquote></details>

Relegated a bit more to Python in 22880a8


<!-- Sent from Reviewable.io -->

sloretz commented:
I think this PR is just waiting on review. There's a little bit of test coverage now in the form of linters, a couple integration tests (one for C++, one for Python) and a small unit test for the `DrakeRos` class.

sloretz commented:
Most of this is going into https://github.com/RobotLocomotion/drake-ros

I'll merge this and archive this repo.
